File: src/084.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/param.h>

#define BOARD_LEN 40
#define TURNS 10000000
#define DICE_SIDES 4

enum Board 
{
    GO   = 0,
    A1   = 1,
    CC1  = 2,
    A2   = 3,
    T1   = 4,
    R1   = 5,
    B1   = 6,
    CH1  = 7,
    B2   = 8,
    B3   = 9,
    JAIL = 10,
    C1   = 11,
    U1   = 12,
    C2   = 13,
    C3   = 14,
    R2   = 15,
    D1   = 16,
    CC2  = 17,
    D2   = 18,
    D3   = 19,
    FP   = 20,
    E1   = 21,
    CH2  = 22,
    E2   = 23,
    E3   = 24,
    R3   = 25,
    F1   = 26,
    F2   = 27,
    U2   = 28,
    F3   = 29,
    G2J  = 30,
    G1   = 31,
    G2   = 32,
    CC3  = 33,
    G3   = 34,
    R4   = 35,
    CH3  = 36,
    H1   = 37,
    T2   = 38,
    H2   = 39
};

int counter[BOARD_LEN];

int compare_squares(const void* a, const void* b) 
{
    int x = counter[*(const int*)a];
    int y = counter[*(const int*)b];
    return y - x;  /* reverse */

}

int main() 
{
    for (int i=0; i<BOARD_LEN; ++i)
        counter[i] = 0;
    int pos = 0;
    int doubles = 0;

    for (int t=0; t<TURNS; ++t)
    {
        /* roll dice */
        int d1 = rand() % DICE_SIDES + 1;
        int d2 = rand() % DICE_SIDES + 1;
        int d = d1 + d2;

        if (d1 == d2) doubles += 1;
        else          doubles = 0;

        //printf("dice %d+%d=%d doubles:%d\n", d1, d2, d, doubles); 

        if (doubles == 3)
        {
            doubles = 0;
            pos = JAIL;
        } 
        else 
            pos = (pos + d) % BOARD_LEN;

        /* handle special squares */
        /* go to jail */
        if (pos == G2J) 
            pos = JAIL;

        /* chance */
        if (pos == CH1 || pos == CH2 || pos == CH3)
        {
            int card = rand() % 16;
            //printf("chance %d\n", card);
            if (card == 0)
                pos = GO;
            else if (card == 1)
                pos = JAIL;
            else if (card == 2)
                pos = C1;
            else if (card == 3)
                pos = E3;
            else if (card == 4)
                pos = H2;
            else if (card == 5)
                pos = R1;
            else if (card == 6 || card == 7)
            {
                while (pos != R1 && pos != R2 && pos != R3)
                    pos = (pos + 1) % BOARD_LEN;
            }
            else if (card == 8)
            {
                while (pos != U1 && pos != U2)
                    pos = (pos + 1) % BOARD_LEN;
            } 
            else if (card == 9) 
                pos = (pos - 3 + BOARD_LEN) % BOARD_LEN;
        }


        /* chest */
        if (pos == CC1 || pos == CC2 || pos == CC3) 
        {
            int card = rand() % 16;
           //printf("chest %d\n", card);
            if (card == 0)
                pos = GO;
            else if (card == 1)
                pos = JAIL;
       }


        /* record position after roll */
        //printf("pos %d\n", pos);
        counter[pos] += 1;

    }

    /* top 3 squares */
    int topsq[BOARD_LEN];
    for (int i=0; i<BOARD_LEN; ++i)
        topsq[i] = i;

    qsort(topsq, BOARD_LEN, sizeof(int), compare_squares);

    for (int i=0; i<3; ++i)
    {
        int sq = topsq[i];
        printf("%d %d\n", sq, counter[sq]);
    }

    printf("%02d%02d%02d\n", topsq[0], topsq[1], topsq[2]);

}

File: src/084.py
from enum import Enum
from random import randint, sample

BOARD = ("GO", "A1", "CC1", "A2", "T1", "R1", "B1", "CH1", "B2", "B3",
      "JAIL", "C1", "U1", "C2", "C3", "R2", "D1", "CC2", "D2", "D3",
      "FP", "E1", "CH2", "E2", "E3", "R3", "F1", "F2", "U2", "F3",
      "G2J", "G1", "G2", "CC3", "G3", "R4", "CH3", "H1", "T2", "H2")

GAMES = 100
TURNS = 1000
counter = [0] * len(BOARD)

for game in range(GAMES):
    print("new game")
    pos = 0
    doubles = 0
    chest = sample(range(16), k=16)
    chest_ind = 0
    chance = sample(range(16), k=16)
    chance_ind = 0
    print("chest", chest)
    print("chance", chance)

    for turn in range(TURNS):
        # roll dice
        d1 = randint(1, 6)
        d2 = randint(1, 6)
        d = d1 + d2

        if d1 == d2:
            doubles += 1
        else:
            doubles = 0

        print(f"dice {d1}+{d2}={d} doubles:{doubles}")


        if doubles == 3:
            print("3 doubles! go to jail!")
            doubles = 0
            pos = BOARD.index("JAIL")

        else:
            # make move
            pos = (pos + d) % len(BOARD)

        # G2J square
        if pos == BOARD.index("G2J"):
            pos = BOARD.index("JAIL")

        # community chest card
        if BOARD[pos].startswith("CC"):
            print("chest", chest_ind)
            if chest[chest_ind] == 0:
                pos = BOARD.index("GO")
            elif chest[chest_ind] == 1:
                pos = BOARD.index("JAIL")

            chest_ind = (chest_ind + 1) % len(chest)

        # chance card
        elif BOARD[pos].startswith("CH"):
            card = chance[chance_ind]
            print("chance", chance_ind, card)
            if card == 0:
                pos = BOARD.index("GO")
            elif card == 1:
                pos = BOARD.index("JAIL")
            elif card == 2:
                pos = BOARD.index("C1")
            elif card == 3:
                pos = BOARD.index("E3")
            elif card == 4:
                pos = BOARD.index("H2")
            elif card == 5:
                pos = BOARD.index("R1")
            elif card in (6, 7):
                while not BOARD[pos].startswith("R"):
                    pos = (pos + 1) % len(BOARD)
            elif card == 8:
                while not BOARD[pos].startswith("U"):
                    pos = (pos + 1) % len(BOARD)
            elif card == 9:
                pos = (pos - 3) % len(BOARD)

            chance_ind = (chance_ind + 1) % len(chance)

        # record position after roll
        print("position", pos, BOARD[pos])
        counter[pos] += 1

print("counter", counter)

most_visited = sorted(range(len(BOARD)),
                      key=lambda i:counter[i],
                      reverse=True)

for i in range(3):
    sq = most_visited[i]
    print(BOARD[sq], counter[sq])